### PR TITLE
CSI - API Changes for Pod Inline Volume Support

### PR DESCRIPTION
**Pull request by vladimirvivien**
**What this PR does / why we need it**:
This PR introduces the changes needed in the API to support inline volume for pods using CSI drivers.  Detail can be found in feature https://github.com/kubernetes/features/issues/596

```release-note
CSI support for inline ephemeral volumes that can be specified in pod specs.
```

⚠️ The will no longer track the implementation changes for CSI Inline Volume.  Its content has been combined with PR https://github.com/kubernetes/kubernetes/pull/68232 (this PR has cherry-picked there).  Follow https://github.com/kubernetes/kubernetes/pull/68232 for all future changes. ⚠️ 

/sig storage

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -1556,6 +1559,57 @@ type CSIPersistentVolumeSource struct {
...
+
+	// VolumeHandle is the unique volume name returned by the CSI volume
+	// plugin’s CreateVolume to refer to the volume on all subsequent calls.
+	// Required.
+	VolumeHandle string
+
+	// Optional: The value to pass to ControllerPublishVolumeRequest.
+	// Defaults to false (read/write).
+	// +optional
+	ReadOnly bool
```
Comment (thockin) [inline pkg/apis/core/types.go]: All optional fields should be pointers or nil-able types (maps, slices)

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -1556,6 +1559,57 @@ type CSIPersistentVolumeSource struct {
...
+
+	// Filesystem type to mount.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs".
+	// +optional
+	FSType string
+
+	// Attributes of the volume to publish.
```
Comment (thockin) [inline pkg/apis/core/types.go]: Can you expand this?  What is done with them?

File: pkg/apis/core/types.go [hunk 3]
```diff
@@ -1556,6 +1559,57 @@ type CSIPersistentVolumeSource struct {
...
+
+	// Optional: The value to pass to ControllerPublishVolumeRequest.
+	// Defaults to false (read/write).
+	// +optional
+	ReadOnly bool
+
+	// Filesystem type to mount.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs".
+	// +optional
```
Comment (thockin) [inline pkg/apis/core/types.go]: "Default is auto-detect.  If not formatted the FS will be chosen by the node." ?

File: pkg/apis/core/types.go [hunk 4]
```diff
@@ -1556,6 +1559,57 @@ type CSIPersistentVolumeSource struct {
 	NodePublishSecretRef *SecretReference
 }
 
+// Represents a source location of a volume to mount, managed by an external CSI driver
+type CSIVolumeSource struct {
+	// Driver is the name of the driver to use for this volume.
+	// Required.
+	Driver string
```
Comment (thockin) [inline pkg/apis/core/types.go]: do we need to add quota (ok as followup) on this?

**Review by thockin — COMMENTED** [review]
Small things

File: pkg/apis/core/types.go (hunk 2 above)
Comment (vladimirvivien) [reply pkg/apis/core/types.go]: Will update the comment.

File: pkg/apis/core/types.go (hunk 3 above)
Comment (vladimirvivien) [reply pkg/apis/core/types.go]: > "Default is auto-detect. If not formatted the FS will be chosen by the node."

I see this quoted, Where is this from ?
Also,  for CSI, fstype is passed as-is delegating its interpretation to the CSI driver.

**Review by vladimirvivien — COMMENTED** [review]
Replies to Tim comments.

File: pkg/apis/core/types.go (hunk 3 above)
Comment (thockin) [reply pkg/apis/core/types.go]: I meant you need to document what the default is when a user doesn't specify, and that the default shouldn't be defined as a particular FS but as "the system will choose for you".

**Comment by thockin** [comment]
I'm trying to keep track of PRs that need re-review.  This one has unresolved comments.

**Comment by saad-ali** [comment]
FYI @vladimirvivien #67803 has merged. Feel free to rebase. That PR did not include the generateVolumeHandle API. We can do that in a follow up PR (and slip it to next release, if needed).

File: pkg/apis/core/types.go (hunk 3 above)
Comment (vladimirvivien) [reply pkg/apis/core/types.go]: Will fix

**Comment by vladimirvivien** [comment]
@thockin I addressed the comments.

File: pkg/apis/core/types.go [hunk 5]
```diff
@@ -1570,6 +1573,59 @@ type CSIPersistentVolumeSource struct {
...
+	// +optional
+	VolumeAttributes map[string]string
+
+	// ControllerPublishSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// ControllerPublishVolume and ControllerUnpublishVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secrets are passed.
+	// +optional
+	ControllerPublishSecretRef *SecretReference
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/core/types.go]: I'd expect references inside a namespaced object to only be in that namespace

**Comment by vladimirvivien** [comment]
@liggitt @thockin @saad-ali  PTAL

File: pkg/apis/core/types.go [hunk 6]
```diff
@@ -1570,6 +1573,59 @@ type CSIPersistentVolumeSource struct {
 	NodePublishSecretRef *SecretReference
 }
 
+// Represents a source location of a volume to mount, managed by an external CSI driver
+type CSIVolumeSource struct {
```
[the remark this answers is not in the document]
Comment (saad-ali) [reply pkg/apis/core/types.go]: Ok I see it: https://github.com/kubernetes/kubernetes/pull/68232

File: pkg/apis/policy/types.go [hunk 7]
```diff
@@ -232,6 +232,10 @@ type PodSecurityPolicySpec struct {
 	// Empty or nil indicates that only the DefaultProcMountType may be used.
 	// +optional
 	AllowedProcMountTypes []api.ProcMountType
+	// AllowedCSIDrivers is a whitelist of allowed CSI drivers.  This slice will be checked
```
Comment (msau42) [inline pkg/apis/policy/types.go]: Should this clarify that it's only limiting inline csi drivers (and not pv.csi drivers)?

**Review by msau42 — CHANGES_REQUESTED** [review]
Since this is an alpha feature, it needs a feature gate, and the api fields need to be dropped when the feature gate is not enabled.

See https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#alpha-field-in-existing-api-version

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/67452/commits/b16725d8a6bb8d2615cfb2203642869c5d47478f#r214546328 is still outstanding, right?

File: pkg/apis/core/types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/core/types.go]: other inline volume sources (like ScaleIOVolumeSource, StorageOSVolumeSource, CephFSVolumeSource, etc) use `LocalObjectReference` for this, where the namespace is implicitly the namespace of the pod

**Comment by vladimirvivien** [comment]
@liggitt yes, this feature is still outstanding. Looking for approval.

File: pkg/apis/policy/types.go (hunk 7 above)
Comment (vladimirvivien) [reply pkg/apis/policy/types.go]: This will be a way to whitelist any CSI driver, not just inline.

**Comment by vladimirvivien** [comment]
@liggit I see what you meant.  The implementation code is forcing the namespace right now, but I will update with LocalObjectRef to have that done implicitly.

File: pkg/apis/core/types.go [hunk 8]
```diff
@@ -1574,6 +1577,59 @@ type CSIPersistentVolumeSource struct {
...
+	ReadOnly *bool
+
+	// Filesystem type to mount. Ex. "ext4", "xfs", "ntfs".
+	// If not provided, the empty value is passed to the associated CSI driver
+	// which will determine the default filesystem to apply.
+	//
+	// +optional
+	FSType string
+
+	// VolumeAttributes store immutable propoerites of the volume copied during provision.
```
Comment (liggitt) [inline pkg/apis/core/types.go]: typo: propoerites

File: pkg/apis/core/types.go [hunk 9]
```diff
@@ -1574,6 +1577,59 @@ type CSIPersistentVolumeSource struct {
...
+	// These attributes are passed back to the driver during controller publish calls.
+	// +optional
+	VolumeAttributes map[string]string
+
+	// ControllerPublishSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// ControllerPublishVolume and ControllerUnpublishVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secrets are passed.
```
Comment (liggitt) [inline pkg/apis/core/types.go]: "If the secret object contains more than one secret, all secrets are passed." should probably be "All keys in the referenced secret object are passed to the CSI driver."

File: pkg/apis/core/validation/validation.go [hunk 10]
```diff
@@ -1491,6 +1501,15 @@ func validateCSIPersistentVolumeSource(csi *core.CSIPersistentVolumeSource, fldP
 	return allErrs
 }
 
+func validateCSIVolumeSource(csi *core.CSIVolumeSource, fldPath *field.Path) field.ErrorList {
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go]: validation seems underspecified
* should driver name be validated (max length, allowable format)
* volume handle max length, allowable format
* the API doc indicates VolumeHandle is required. is that where the discussion about per-pod volumes ended up? either validate it as required here or fix the API doc
* if secret references are non-nil, require name to be present and a well-formed secret name

File: pkg/apis/policy/types.go (hunk 7 above)
Comment (msau42) [reply pkg/apis/policy/types.go]: If that's the case, then I think we may need two separate fields, one for inline and one for PV.  For example, we want to allow gce PD CSI driver as a PV, but disallow it as inline because a lot of functionality won't work inline.

Comment (liggitt) [reply pkg/apis/policy/types.go]: whether it will work or not is not the point of PSP. you can specify driver names that don't even exist in a CSI PV today, and it fails at the point of calling the driver, correct?

PSP is about explicitly whitelisting which drivers a user is allowed to reference

File: pkg/apis/policy/types.go [hunk 11]
```diff
@@ -304,6 +308,12 @@ type AllowedFlexVolume struct {
 	Driver string
 }
 
+// AllowedCSIDriver represents a single CSI Driver that is allowed to be used.
+type AllowedCSIDriver struct {
+	// Name of the CSI driver
+	Name string
```
Comment (liggitt) [inline pkg/apis/policy/types.go]: not seeing any validation of this... Name should be required, and ensured to conform to well-formed CSI driver names (valid length and characters)

File: pkg/apis/policy/types.go [hunk 12]
```diff
@@ -232,6 +232,10 @@ type PodSecurityPolicySpec struct {
 	// Empty or nil indicates that only the DefaultProcMountType may be used.
 	// +optional
 	AllowedProcMountTypes []api.ProcMountType
+	// AllowedCSIDrivers is a whitelist of allowed CSI drivers.  This slice will be checked
+	// to verify which CSI driver is allowed to run in the cluster.
+	// +optional
+	AllowedCSIDrivers []AllowedCSIDriver
```
Comment (liggitt) [inline pkg/apis/policy/types.go]: move this just under AllowedFlexVolumes

File: pkg/apis/core/types.go (hunk 9 above)
Comment (vladimirvivien) [reply pkg/apis/core/types.go]: Got it!

File: pkg/apis/core/validation/validation.go (hunk 10 above)
Comment (vladimirvivien) [reply pkg/apis/core/validation/validation.go]: @liggitt 
> the API doc indicates VolumeHandle is required. is that where the discussion about per-pod volumes ended up? either validate it as required here or fix the API doc
I will fix the API doc.  Where we ended up is that it will not be required for inline per-vol pod.

File: pkg/apis/policy/types.go (hunk 12 above)
Comment (vladimirvivien) [reply pkg/apis/policy/types.go]: Ok will do.

File: pkg/apis/core/validation/validation.go (hunk 10 above)
Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: sounds good. be sure to add `+optional` and `omitempty` as appropriate

File: pkg/apis/storage/validation/validation.go [hunk 13]
```diff
@@ -164,8 +164,37 @@ func validateAttacher(attacher string, fldPath *field.Path) field.ErrorList {
 // validateSource tests if the source is valid for VolumeAttachment.
 func validateVolumeAttachmentSource(source *storage.VolumeAttachmentSource, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if source.PersistentVolumeName == nil || len(*source.PersistentVolumeName) == 0 {
```
Comment (liggitt) [inline pkg/apis/storage/validation/validation.go]: this field was previously marked as optional, but required by validation. that means an n-1 apiserver would see a persisted object with `inlineVolumeSource` specified as invalid (it wouldn't know about `inlineVolumeSource` and would think the volumeAttachment had not specified any source)

need to think about how to make the addition of this field safe for rollback

File: pkg/apis/storage/validation/validation.go [hunk 14]
```diff
@@ -164,8 +164,37 @@ func validateAttacher(attacher string, fldPath *field.Path) field.ErrorList {
...
+		if source.PersistentVolumeName != nil && source.InlineVolumeSource != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, "source", "either persistentVolumeName or inlineVolumeSource can be provided"))
+		}
+	}
+
+	if (source.PersistentVolumeName != nil && len(*source.PersistentVolumeName) == 0) && source.InlineVolumeSource == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, "persistentVolumeName", "persistentVolumeName cannot be empty"))
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
```
Comment (liggitt) [inline pkg/apis/storage/validation/validation.go]: this is pretty hard to follow... can we gather the `CSIInlineVolume`-protected bits in a single block?

File: pkg/apis/storage/validation/validation.go [hunk 15]
```diff
@@ -164,8 +164,37 @@ func validateAttacher(attacher string, fldPath *field.Path) field.ErrorList {
...
+
+	return allErrs
+}
+
+// validateSource tests if the source is valid for VolumeAttachment.
+func validateVolumeAttachmentInlineVolumeSource(source *storage.InlineVolumeSource, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if source.VolumeSource.CSI == nil {
+		allErrs = append(allErrs, field.Required(fldPath, "volumeSource.CSI"))
+	}
```
Comment (liggitt) [inline pkg/apis/storage/validation/validation.go]: shouldn't this be calling validateVolumeSource(source.VolumeSource) as well? nothing is checking that this source is well-formed

File: pkg/apis/storage/validation/validation_test.go [hunk 16]
```diff
@@ -314,8 +332,46 @@ func TestVolumeAttachmentValidation(t *testing.T) {
...
+				NodeName: "node",
+				Source: storage.VolumeAttachmentSource{
+					InlineVolumeSource: &storage.InlineVolumeSource{},
+				},
+			},
+		},
 	}
 
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+		utilfeature.DefaultFeatureGate.Set("CSIInlineVolume=true")
```
Comment (liggitt) [inline pkg/apis/storage/validation/validation_test.go]: this changes global state and leaves it changed after the test completes. should pass in feature enablement into the validation method or use the `SetFeatureGateDuringTest` helper that restores original state once the test completes

**Comment by liggitt** [comment]
I'm not seeing the new fields (in pods and volumeattachments) getting cleared if the feature gate is disabled. a server without this feature turned on should behave as though the API fields do not exist

File: pkg/apis/policy/types.go (hunk 7 above)
Comment (vladimirvivien) [reply pkg/apis/policy/types.go]: @msau42 to reduce complexity, the meaning of the PSP is simply a list of allowed CSI drivers here. Capabilities such as inline/attached etc is exposed in the `CSIInfo` object.  We can refine its interpretation in upcoming release.

Comment (msau42) [reply pkg/apis/policy/types.go]: Is PSP validated at admission time and/or runtime?  I thought that today we cannot have PV validation in PSP because of cross-object validation and the PV may not exist yet

Comment (liggitt) [reply pkg/apis/policy/types.go]: Admission time. PSP only applies to pods, so this would only restrict inline CSI volume sources

File: pkg/apis/storage/validation/validation.go (hunk 13 above)
Comment (vladimirvivien) [reply pkg/apis/storage/validation/validation.go]: @liggitt in the case of rollback, where `inlineVolumeSource`  is present, but not `persistentVolumeName`, the validation would fail.  I am wrapping the validation in a feature-gate condition forcing an explicit enabling of this feature.

File: pkg/apis/storage/validation/validation.go (hunk 14 above)
Comment (vladimirvivien) [reply pkg/apis/storage/validation/validation.go]: yes

File: pkg/apis/storage/validation/validation.go (hunk 15 above)
Comment (vladimirvivien) [reply pkg/apis/storage/validation/validation.go]: fixed

File: pkg/apis/storage/validation/validation_test.go (hunk 16 above)
Comment (vladimirvivien) [reply pkg/apis/storage/validation/validation_test.go]: added code to turn/off tests

**Comment by vladimirvivien** [comment]
@liggitt addressed all code comments.

**Comment by vladimirvivien** [comment]
@liggitt PTAL. I addressed your concerns.

File: docs/admin/kube-proxy.md [hunk 17]
```diff
@@ -1,3 +1,125 @@
-This file is autogenerated, but we've stopped checking such files into the
-repository to reduce the need for rebases. Please run hack/generate-docs.sh to
-populate this file.
+## kube-proxy
```
Comment (thockin) [inline docs/admin/kube-proxy.md]: These changes should not be committed

File: pkg/api/pod/util.go [hunk 18]
```diff
@@ -326,3 +336,13 @@ func DropDisabledVolumeDevicesAlphaFields(podSpec *api.PodSpec) {
 		}
 	}
 }
+
+// DropDisabledCSIVolumeSourceAlphaFields removes disabled alpha fields from []CSIVolumeSource.
+// This should be called from PrepareForCreate/PrepareForUpdate for all pod specs resources containing a CSIVolumeSource
+func DropDisabledCSIVolumeSourceAlphaFields(podSpec *api.PodSpec) {
```
Comment (thockin) [inline pkg/api/pod/util.go]: @liggitt I have seen some PRs that drop alpha fields and others that fail validation if they are set.  Do we have a rule on this?

File: pkg/apis/core/types.go [hunk 19]
```diff
@@ -1574,6 +1577,61 @@ type CSIPersistentVolumeSource struct {
...
+	// Defaults to false (read/write).
+	// +optional
+	ReadOnly *bool
+
+	// Filesystem type to mount. Ex. "ext4", "xfs", "ntfs".
+	// If not provided, the empty value is passed to the associated CSI driver
+	// which will determine the default filesystem to apply.
+	//
+	// +optional
+	FSType string
```
Comment (thockin) [inline pkg/apis/core/types.go]: All optional fields should be pointers or nilable

File: pkg/api/pod/util.go (hunk 18 above)
Comment (liggitt) [reply pkg/api/pod/util.go]: >  I have seen some PRs that drop alpha fields and others that fail validation if they are set. Do we have a rule on this?

The guidelines in https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#adding-unstable-features-to-stable-versions call for both. The validation failure is redundant if the pre-create/update drops are done, but is a belt/suspenders check to ensure no alpha data is allowed.

File: docs/admin/kube-proxy.md (hunk 17 above)
Comment (vladimirvivien) [reply docs/admin/kube-proxy.md]: Not sure why the update-generated-docs.sh script did that.  The verify- counter part was failing. Ok will remove.

File: pkg/apis/core/types.go (hunk 19 above)
Comment (vladimirvivien) [reply pkg/apis/core/types.go]: @thockin ok I will make all optional pointers or nilable.

**Comment by vladimirvivien** [comment]
@liggitt @thockin - all cleaned up.

**Comment by vladimirvivien** [comment]
/assign @liggitt

PTAL

**Comment by liggitt** [comment]
I'm actually out of the office. I can review when I get back (which will be prior to master opening for 1.13)

**Comment by vladimirvivien** [comment]
@lavalamp PTAL

**Comment by vladimirvivien** [comment]
@lavalamp @liggitt @saad-ali @msau42 PTAL

File: pkg/apis/core/validation/validation.go [hunk 20]
```diff
@@ -638,6 +638,16 @@ func validateVolumeSource(source *core.VolumeSource, fldPath *field.Path, volNam
...
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIInlineVolume) {
+		if source.CSI != nil {
+			if numVolumes > 0 {
+				allErrs = append(allErrs, field.Forbidden(fldPath.Child("csi"), "may not specify more than 1 volume type"))
+			} else {
+				numVolumes++
+				allErrs = append(allErrs, validateCSIVolumeSource(source.CSI, fldPath.Child("csi"))...)
+			}
+		}
+	}
```
Comment (saad-ali) [inline pkg/apis/core/validation/validation.go]: If feature is disabled, should validation reject the object? Accepting it without validation seems like a bad idea.

Comment (vladimirvivien) [reply pkg/apis/core/validation/validation.go]: @saad-ali not quite following you. 

If feature is enabled, only then the object is validated because when the feature is not enabled, the object (`VolumeSource.CSI`) is cleared as nil (see https://github.com/kubernetes/kubernetes/blob/19be94c88bec6b3cd5983e3f7a468b1e7b30c8bb/pkg/api/pod/util.go#L342)

Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: that is correct, but as described in https://github.com/kubernetes/kubernetes/pull/67452#discussion_r215694746, also checking in validation ensured that if the PrepareForCreate/PrepareForUpdate check is removed or skipped, validation prevents alpha data from being persisted when the feature is not enabled.

Comment (msau42) [reply pkg/apis/core/validation/validation.go]: Just in case the field drop doesn't work, it would be good to return an error in validation if CSI is also set and the feature is off.

Comment (vladimirvivien) [reply pkg/apis/core/validation/validation.go]: @liggitt I misread @saad-ali concerns.  Fixed cc @msau42 

PTAL

**Comment by vladimirvivien** [comment]
@lavalamp @liggitt @saad-ali @msau42 PTAL

**Comment by vladimirvivien** [comment]
@thockin @saad-ali @liggitt PTAL.  Anything else this PR needs?

File: pkg/apis/storage/types.go [hunk 21]
```diff
@@ -157,7 +157,22 @@ type VolumeAttachmentSource struct {
...
+	// VolumeSource is copied from the pod. It ensures that attacher has enough
+	// information to detach a volume when the pod is deleted before detaching.
+	// Only CSIVolumeSource can be set.
+	// Required.
+	VolumeSource api.VolumeSource
+
+	// Namespace of the pod with in-line volume. It is used to resolve
+	// references to Secrets in VolumeSource.
+	// Required.
+	Namespace string
```
Comment (saad-ali) [inline pkg/apis/storage/types.go]: Sorry if I missed the conversation around this. But just to verify, specifying the namespace here is considered safe because `VolumeAttachment` should only be created by `kube-controller-manager` right? CC @liggitt

**Review by saad-ali — COMMENTED** [review]
LGTM

File: pkg/apis/storage/types.go (hunk 21 above)
Comment (liggitt) [reply pkg/apis/storage/types.go]: > just to verify, specifying the namespace here is considered safe because VolumeAttachment should only be created by kube-controller-manager right? CC @liggitt

Yes, InlineVolumeSource is only used in VolumeAttachmentSource inside VolumeAttachment, which is a cluster-scoped object.

It contains a denormalized copy of the VolumeSource field from a pod, along with the recorded namespace of the pod (since all the resource references inside the v1.VolumeSource implicitly refer to the namespace of the pod, but that information is not available when the volume source is copied up the a cluster-level object).

File: pkg/apis/core/types.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (thockin) [reply pkg/apis/core/types.go]: Hey guys.  So, I think this is a pattern we need to do less of.  Getting an API in with no implementation is always going to be riskier than putting them together.  Getting a review on API before implementing is fine, but I think we want to merge them together.

File: pkg/apis/policy/validation/validation.go [hunk 22]
```diff
@@ -39,6 +40,13 @@ import (
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 )
 
+const (
+	csiDriverNameRexpFmt    = `^[a-zA-Z0-9][-a-zA-Z0-9_.]{0,61}[a-zA-Z-0-9]$`
```
Comment (thockin) [inline pkg/apis/policy/validation/validation.go]: This requires at least 2 letters - is that the correct spec?  Or is it something like 
 
`^[a-zA-Z0-9]([-a-zA-Z0-9_.]{0,61}[a-zA-Z-0-9])?$`

File: pkg/apis/storage/validation/validation.go (hunk 13 above)
Comment (thockin) [reply pkg/apis/storage/validation/validation.go]: Strictly speaking there's no way to convert a field from required to optional and be rollback-safe.  This is kind of a problem.  Any creative ideas here?

**Comment by vladimirvivien** [comment]
@thockin Thanks for continuing to review.  As far as implementation, I do have another PR (based on this one) with the implementation.  If that is the preferred way of moving forward, I don't see a problem - 

https://github.com/kubernetes/kubernetes/pull/68232

I will mark this as paused and redirect future reviews (above).

**Comment by vladimirvivien** [comment]
⚠️ This PR will no longer track the implementation for inline volume.  It will be closed.  Its content has been rebased and combined to PR https://github.com/kubernetes/kubernetes/pull/68232 for future changes. ⚠️

**Comment by spiffxp** [comment]
/close
closing per the above comment
